doc/memory_statistics/memory_statistics_hld.md

```diff
@@ -0,0 +1,373 @@
+
+**MEMORY_STATS Configuration Parameters**
+
+| Parameter           | Type        | Description                                                    |
+|---------------------|-------------|----------------------------------------------------------------|
+| enabled             | boolean     | Enable or disable memory statistics collection.                |
+| sampling_interval   | unit16      | Interval for memory data collection.       		     |
+| retention_period    | unit16      | Duration for which memory data is retained.		     |   
+
+**Config DB Schema**
+```json
+
+MEMORY_STATS_TABLE: {
+    "MemoryStats": {
+        "enabled": "true",
+        "sampling_interval": "5",
+        "retention_period":  "15"
+    }
+}
+``` 
+
+
+## Configuration and Management
+
+### CLI/YANG Model Enhancements
+
+#### CLI Commands
+
+**Set the Frequency of Memory Data Collection**
+
+To configure the interval for memory data collection, use the following command:
+
+	admin@sonic:~$ config memory-stats sampling-interval <interval>
+ Default sampling-interval is 5 minutes
+
+**Adjust the Data Retention Period**
+
+To set how long the memory data should be retained, use the following command:
+
+	admin@sonic:~$ config memory-stats retention-period <period>
+ Default retention-period is 15 days
+
+**Enable/Disable Memory Statistics Monitoring**
+
+To enable or disable the memory statistics monitoring feature, use the following command:
+
+	admin@sonic:~$ config memory-stats enable/disable
+ By default, it is disabled.
+
+**View Memory Usage**
+
+To display memory usage statistics, use the following command with optional parameters for time range and specific metrics:
+
+	admin@sonic:~$ show memory-stats [--from <date-time>] [--to <date-time>] [--select <metric>]
+
+**Command Definition**
+  - **show memory-stats:** Display basic memory usage statistics
+  - **--from <date-time>:** Display memory statistics from the specified start date-time.
+  - **--to <date-time>:** Display memory statistics up to the specified end date-time.
+  - **--select <metric>:** Display specific memory statistics, such as total memory.
+
+ **Sample Output for Memory Usage**
+
+Below is an example of the Memory Statistics output as it appears in the CLI. This display provides a summary of system memory metrics over a default time period, ideal for routine monitoring and analysis:
+
+
+    admin@sonic:~$ show memory-stats 
+
+
+    Codes:	M - minutes, H - hours, D - days
+	--------------------------------------------------------------------------------
+	Report Generated:    2024-06-15 09:00:00
+    Analysis Period:     From 2024-06-01 09:00:00 to 2024-06-15 09:00:00
+    Interval:	         2 days
+	--------------------------------------------------------------------------------
+	  Metric            Current     High       Low        D1-D3      D3-D5      D5-D7      D7-D9      D9-D11     D11-D13    D13-D15    
```

Comment:
sonic-gnmi (telemetry) already has a comprehensive implementation https://github.com/sonic-net/sonic-gnmi/blob/64ed32b715e1673c3376701ee192d6292f5ec0ec/sonic_data_client/non_db_client.go#L235. Please check and ensure matrix are aligned.
Even better, suggest reuse:
1. delegate telemetry container as the memory collector
2. reduce this HLD scope to only memory-stats and storage, connecting to telemetry container by gnmi protocol
3. ensure this design could run inside and outside sonic switch
4. ensure this feature could be completely disabled inside a sonic switch and no performance burden if disabled

Reply:
- Thank you for the suggestion. We will review the sonic-gnmi implementation to align our metrics with the existing telemetry framework for system consistency. We will also explore reusing the telemetry container for memory collection and narrow the HLD scope to focus on memory statistics and storage.
- Currently, we use the psutil.virtual_memory() function from the psutil library to gather memory data, which is similar to the getProcMeminfo() function in non_db_client.go, as both methods pull information from /proc/meminfo. We believe that psutil offers a more efficient and Pythonic approach compared to relying solely on the telemetry container for this purpose.
- Can you elaborate more on how the design could run inside and outside the sonic switch.

Reply:
if this design rely on telemetry container as the memory stats collector, it can run inside or outside sonic switch.

Reply:
this thread is still applicable.

Reply:
To provide some context on the current design, we gather memory statistics directly from SONiC using the psutil library. The collected data is stored as gzip files in the /var/log directory and is accessible via show commands through Unix sockets.

We appreciate your suggestion to leverage the telemetry container. As part of future enhancements for this feature, we will certainly consider integrating with the telemetry container and aligning the design with your recommendations.

Thank you.